RoleBasedAccessControl sample does not compile

# Description

The role-based access control sample doesn't compile.

1. `git clone https://github.com/Azure/azure-service-bus/tree/master/samples/DotNet/Microsoft.ServiceBus.Messaging/RoleBasedAccessControl`
1. `cd azure-service-bus\samples\DotNet\Microsoft.ServiceBus.Messaging\RoleBasedAccessControl`
1. `msbuild`

## Actual Behavior

Build fails.

Output:

    Build FAILED.

    "<SomePath>\azure-service-bus\samples\DotNet\Microsoft.ServiceBus.Messaging\RoleBasedAccessControl\RoleBasedAccessControl.sln" (default target) (1) ->
    "<SomePath>\azure-service-bus\samples\DotNet\Microsoft.ServiceBus.Messaging\RoleBasedAccessControl\RoleBasedAccessControl.csproj" (default target) (2) ->
    (CoreCompile target) ->
    Program.cs(29,21): error CS0234: The type or namespace name 'Identity' does not exist in the namespace 'Microsoft' (are you missing an assembly reference?) [<SomePath>\azure-service-bus\samples\DotNet\Microsoft.ServiceBus.Messaging\RoleBasedAccessControl\RoleBasedAccessControl.csproj]
    Program.cs(31,27): error CS0234: The type or namespace name 'ServiceBus' does not exist in the namespace 'Microsoft.Azure' (are you missing an assembly reference?) [<SomePath>\azure-service-bus\samples\DotNet\Microsoft.ServiceBus.Messaging\RoleBasedAccessControl\RoleBasedAccessControl.csproj]
    Program.cs(32,27): error CS0234: The type or namespace name 'ServiceBus' does not exist in the namespace 'Microsoft.Azure' (are you missing an assembly reference?) [<SomePath>\azure-service-bus\samples\DotNet\Microsoft.ServiceBus.Messaging\RoleBasedAccessControl\RoleBasedAccessControl.csproj]
    Program.cs(33,27): error CS0234: The type or namespace name 'ServiceBus' does not exist in the namespace 'Microsoft.Azure' (are you missing an assembly reference?) [<SomePath>\azure-service-bus\samples\DotNet\Microsoft.ServiceBus.Messaging\RoleBasedAccessControl\RoleBasedAccessControl.csproj]
    Program.cs(85,37): error CS0246: The type or namespace name 'QueueClient' could not be found (are you missing a using directive or an assembly reference?) [<SomePath>\azure-service-bus\samples\DotNet\Microsoft.ServiceBus.Messaging\RoleBasedAccessControl\RoleBasedAccessControl.csproj]
    Program.cs(38,9): error CS0246: The type or namespace name 'QueueClient' could not be found (are you missing a using directive or an assembly reference?) [<SomePath>\azure-service-bus\samples\DotNet\Microsoft.ServiceBus.Messaging\RoleBasedAccessControl\RoleBasedAccessControl.csproj]
    Program.cs(39,9): error CS0246: The type or namespace name 'QueueClient' could not be found (are you missing a using directive or an assembly reference?) [<SomePath>\azure-service-bus\samples\DotNet\Microsoft.ServiceBus.Messaging\RoleBasedAccessControl\RoleBasedAccessControl.csproj]

## Expected Behavior

Build succeeds.

--- a/samples/DotNet/Microsoft.Azure.ServiceBus/RoleBasedAccessControl/App.config
+++ b/samples/DotNet/Microsoft.Azure.ServiceBus/RoleBasedAccessControl/App.config
@@ -1,0 +1,15 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+  <appSettings>
+    <add key="tenantId" value="" />
+    <add key="clientId" value="" />
+    <add key="serviceBusNamespaceFQDN" value="" /> <!-- e.g. yournamespaceName.servicebus.windows.net -->
+    <add key="queueName" value="" />
+    <add key="thumbPrint" value="" />
+    <add key="clientSecret" value="" />
+    <add key="userName" value="" />
+    <add key="password" value="" />
+    <add key="redirectURI" value=""/>
+      <!-- Make sure this matches the redirect URI you set in AAD in your app. It doesn't need to be the right URL E.g.: http://servicebus.microsoft.com -->
+  </appSettings>
+ </configuration>

--- a/samples/DotNet/Microsoft.Azure.ServiceBus/RoleBasedAccessControl/Program.cs
+++ b/samples/DotNet/Microsoft.Azure.ServiceBus/RoleBasedAccessControl/Program.cs
@@ -1,0 +1,298 @@
+﻿//   
+//   Copyright © Microsoft Corporation, All Rights Reserved
+// 
+//   Licensed under the Apache License, Version 2.0 (the "License"); 
+//   you may not use this file except in compliance with the License. 
+//   You may obtain a copy of the License at
+// 
+//   http://www.apache.org/licenses/LICENSE-2.0 
+// 
+//   THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+//   OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION
+//   ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A
+//   PARTICULAR PURPOSE, MERCHANTABILITY OR NON-INFRINGEMENT.
+// 
+//   See the Apache License, Version 2.0 for the specific language
+//   governing permissions and limitations under the License. 
+
+namespace MessagingSamples
+{
+    using Newtonsoft.Json;
+    using System;
+    using System.Collections.Generic;
+    using System.Configuration;
+    using System.IO;
+    using System.Security.Cryptography.X509Certificates;
+    using System.Text;
+    using System.Threading.Tasks;
+    using Microsoft.Identity.Client;
+    using Microsoft.Azure.ServiceBus.Primitives;
+    using Microsoft.Azure.ServiceBus;
+    using Microsoft.Azure.ServiceBus.InteropExtensions;
+    
+
+    public class Program
+    {
+        QueueClient sendClient;
+        QueueClient receiveClient;
+
+        static readonly string TenantId = ConfigurationManager.AppSettings["tenantId"];
+        static readonly string ClientId = ConfigurationManager.AppSettings["clientId"];
+        static readonly string ServiceBusNamespace = ConfigurationManager.AppSettings["serviceBusNamespaceFQDN"];
+        static readonly string QueueName = ConfigurationManager.AppSettings["queueName"];
+
+        public async Task Run()
+        {
+            Console.WriteLine("Pick a scenario to run and hit ENTER:");
+            Console.WriteLine("1) ManagedServiceIdentity (must run in an Azure VM or Web Job)");
+            Console.WriteLine("2) Interactive User Login");
+            Console.WriteLine("3) Client secret Login");
+            Console.WriteLine("4) Client credential X.509 certificate");
+
+            int option;
+            var sc = Console.ReadLine();
+            if (int.TryParse(sc, out option))
+            {
+                switch (option)
+                {
+                    case 1:
+                        await ManagedServiceIdentityScenario();
+                        break;
+                    case 2:
+                        await UserInteractiveLoginScenario();
+                        break;
+                    case 3:
+                        await ClientCredentialsScenario();
+                        break;
+                    case 4:
+                        await ClientCredentialsCertScenario();
+                        break;
+                }
+            }
+        }
+
+        async Task ManagedServiceIdentityScenario()
+        {
+            var msiTokenProvider = TokenProvider.CreateManagedIdentityTokenProvider();
+            var qc = new QueueClient(new Uri($"sb://{ServiceBusNamespace}/").ToString(), QueueName, msiTokenProvider);
+
+            await SendReceiveAsync(qc);
+
+        }
+
+        async Task SendReceiveAsync(QueueClient qc)
+        {
+
+            this.receiveClient = qc;
+            this.InitializeReceiver();
+
+            this.sendClient = qc;
+            var sendTask = this.SendMessagesAsync();
+
+            Console.ReadKey();
+
+            // shut down the receiver, which will stop the OnMessageAsync loop
+            await this.receiveClient.CloseAsync();
+
+            // wait for send work to complete if required
+            await sendTask;
+
+            await this.sendClient.CloseAsync();
+
+        }
+
+        async Task SendMessagesAsync()
+        {
+            dynamic data = new[]
+            {
+                new {name = "Einstein", firstName = "Albert"},
+                new {name = "Heisenberg", firstName = "Werner"},
+                new {name = "Curie", firstName = "Marie"},
+                new {name = "Hawking", firstName = "Steven"},
+                new {name = "Newton", firstName = "Isaac"},
+                new {name = "Bohr", firstName = "Niels"},
+                new {name = "Faraday", firstName = "Michael"},
+                new {name = "Galilei", firstName = "Galileo"},
+                new {name = "Kepler", firstName = "Johannes"},
+                new {name = "Kopernikus", firstName = "Nikolaus"}
+            };
+
+
+            for (int i = 0; i < data.Length; i++)
+            {
+                var message = new Message(Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(data[i])))
+                {
+                    ContentType = "application/json",
+                    Label = "Scientist",
+                    MessageId = i.ToString(),
+                    TimeToLive = TimeSpan.FromMinutes(2)
+                };
+
+                await this.sendClient.SendAsync(message);
+                lock (Console.Out)
+                {
+                    Console.ForegroundColor = ConsoleColor.Yellow;
+                    Console.WriteLine("Message sent: Id = {0}", message.MessageId);
+                    Console.ResetColor();
+                }
+            }
+        }
+
+        void InitializeReceiver()
+        {
+            // register the OnMessageAsync callback
+            this.receiveClient.RegisterMessageHandler(
+                async (message, token) =>
+                {
+                    if (message.Label != null &&
+                        message.ContentType != null &&
+                        message.Label.Equals("Scientist", StringComparison.InvariantCultureIgnoreCase) &&
+                        message.ContentType.Equals("application/json", StringComparison.InvariantCultureIgnoreCase))
+                    {
+                        var body = message.Body;
+
+                        dynamic scientist = JsonConvert.DeserializeObject(Encoding.UTF8.GetString(body));
+                        
+                        lock (Console.Out)
+                        {
+                            Console.ForegroundColor = ConsoleColor.Cyan;
+                            Console.WriteLine(
+                                "\t\t\t\tMessage received: \n\t\t\t\t\t\tMessageId = {0}, \n\t\t\t\t\t\tSequenceNumber = {1}, \n\t\t\t\t\t\tEnqueuedTimeUtc = {2}," +
+                                "\n\t\t\t\t\t\tExpiresAtUtc = {5}, \n\t\t\t\t\t\tContentType = \"{3}\", \n\t\t\t\t\t\tSize = {4},  \n\t\t\t\t\t\tContent: [ firstName = {6}, name = {7} ]",
+                                message.MessageId,
+                                message.SystemProperties.SequenceNumber,
+                                message.SystemProperties.EnqueuedTimeUtc,
+                                message.ContentType,
+                                message.Size,
+                                message.ExpiresAtUtc,
+                                scientist.firstName,
+                                scientist.name);
+                            Console.ResetColor();
+                        }
+                    }
+                    await this.receiveClient.CompleteAsync(message.SystemProperties.LockToken);
+                },
+                new MessageHandlerOptions((eventargs) => { return Task.CompletedTask; })
+                {
+                    AutoComplete = false,
+                    MaxConcurrentCalls = 1
+                });
+        }
+
+        async Task UserInteractiveLoginScenario()
+        {
+            var aadTokenProvider = TokenProvider.CreateAzureActiveDirectoryTokenProvider(async (audience, authority, state) =>
+            {
+                var app = PublicClientApplicationBuilder.Create(ClientId)
+                            .WithRedirectUri(ConfigurationManager.AppSettings["redirectURI"])
+                            .Build();
+
+                var serviceBusAudience = new Uri("https://servicebus.azure.net");
+
+                var authResult = await app.AcquireTokenInteractive(new string[] { $"{serviceBusAudience}/.default" }).ExecuteAsync();
+
+                return authResult.AccessToken;
+            }, $"https://login.windows.net/{TenantId}");
+
+            var qc = new QueueClient(new Uri($"sb://{ServiceBusNamespace}/").ToString(), QueueName, aadTokenProvider);
+
+            await SendReceiveAsync(qc);
+
+        }
+
+        async Task ClientCredentialsScenario()
+        {
+            var aadTokenProvider = TokenProvider.CreateAzureActiveDirectoryTokenProvider
+                (async (audience, authority, state) =>
+            {
+                IConfidentialClientApplication app = ConfidentialClientApplicationBuilder.Create(ClientId)
+                                .WithAuthority(authority)
+                                .WithClientSecret(ConfigurationManager.AppSettings["clientSecret"])
+                                .Build();
+
+                var serviceBusAudience = new Uri("https://servicebus.azure.net");
+
+                var authResult = await app.AcquireTokenForClient(new string[] { $"{serviceBusAudience}/.default" }).ExecuteAsync();
+                return authResult.AccessToken;
+
+            }, $"https://login.windows.net/{TenantId}");
+
+            var qc = new QueueClient(new Uri($"sb://{ServiceBusNamespace}/").ToString(), QueueName, aadTokenProvider);
+
+            await SendReceiveAsync(qc);
+        }
+
+        X509Certificate2 GetCertificate()
+        {
+            List<StoreLocation> locations = new List<StoreLocation>
+                {
+                    StoreLocation.CurrentUser,
+                    StoreLocation.LocalMachine
+                };
+
+            foreach (var location in locations)
+            {
+                X509Store store = new X509Store(StoreName.My, location);
+                try
+                {
+                    store.Open(OpenFlags.ReadOnly | OpenFlags.OpenExistingOnly);
+                    X509Certificate2Collection certificates = store.Certificates.Find(
+                        X509FindType.FindByThumbprint, ConfigurationManager.AppSettings["thumbPrint"], true);
+                    if (certificates.Count >= 1)
+                    {
+                        return certificates[0];
+                    }
+                }
+                finally
+                {
+                    store.Close();
+                }
+            }
+
+            throw new ArgumentException($"A Certificate with Thumbprint '{ConfigurationManager.AppSettings["thumbPrint"]}' could not be located.");
+        }
+
+        async Task ClientCredentialsCertScenario()
+        {
+            X509Certificate2 certificate = GetCertificate();
+
+            var aadTokenProvider = TokenProvider.CreateAzureActiveDirectoryTokenProvider(async (audience, authority, state) =>
+            {
+                IConfidentialClientApplication app = ConfidentialClientApplicationBuilder.Create(ClientId)
+                                .WithAuthority(authority)
+                                .WithCertificate(certificate)
+                                .Build();
+
+                var serviceBusAudience = new Uri("https://servicebus.azure.net");
+
+                var authResult = await app.AcquireTokenForClient(new string[] { $"{serviceBusAudience}/.default" }).ExecuteAsync();
+                return authResult.AccessToken;
+
+            }, $"https://login.windows.net/{TenantId}");
+
+            var qc = new QueueClient(new Uri($"sb://{ServiceBusNamespace}/").ToString(), QueueName, aadTokenProvider);
+
+            await SendReceiveAsync(qc);
+
+        }
+
+        public static int Main(string[] args)
+        {
+            try
+            {
+                var app = new Program();
+                app.Run().GetAwaiter().GetResult();
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(e.ToString());
+                Console.ReadLine();
+                return 1;
+            }
+            return 0;
+        }
+
+
+    }
+}
+

--- a/samples/DotNet/Microsoft.Azure.ServiceBus/RoleBasedAccessControl/README.md
+++ b/samples/DotNet/Microsoft.Azure.ServiceBus/RoleBasedAccessControl/README.md
@@ -1,0 +1,3 @@
+# Role based access sample
+
+For more information about Role based access (RBAC), follow [this link](https://docs.microsoft.com/azure/service-bus-messaging/service-bus-role-based-access-control).

--- a/samples/DotNet/Microsoft.Azure.ServiceBus/RoleBasedAccessControl/RoleBasedAccessControl.csproj
+++ b/samples/DotNet/Microsoft.Azure.ServiceBus/RoleBasedAccessControl/RoleBasedAccessControl.csproj
@@ -1,0 +1,18 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFrameworks>net462;netcoreapp3.0</TargetFrameworks>
+
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Azure.ServiceBus" Version="4.1.3" />
+    <PackageReference Include="Microsoft.CSharp" Version="4.7.0" />
+    <PackageReference Include="Microsoft.Identity.Client" Version="4.16.1" />
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
+    <PackageReference Include="System.Configuration.ConfigurationManager" Version="4.7.0" />
+    <PackageReference Include="System.Diagnostics.Debug" Version="4.3.0" />
+    <PackageReference Include="Microsoft.Azure.Services.AppAuthentication" version="1.5.0" />
+  </ItemGroup>
+
+</Project>

--- a/samples/DotNet/Microsoft.Azure.ServiceBus/RoleBasedAccessControl/RoleBasedAccessControl.sln
+++ b/samples/DotNet/Microsoft.Azure.ServiceBus/RoleBasedAccessControl/RoleBasedAccessControl.sln
@@ -1,0 +1,26 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio 14
+VisualStudioVersion = 14.0.24720.0
+MinimumVisualStudioVersion = 10.0.40219.1
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution Items", "{84D60567-DAD4-47CF-B0E2-B9993BC76171}"
+	ProjectSection(SolutionItems) = preProject
+		README.md = README.md
+	EndProjectSection
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "RoleBasedAccessControl", "RoleBasedAccessControl.csproj", "{247F0154-3F6B-409B-BEBA-1BF9302D3B9E}"
+EndProject
+Global
+	GlobalSection(SolutionConfigurationPlatforms) = preSolution
+		Debug|AnyCPU = Debug|AnyCPU
+		Release|AnyCPU = Release|AnyCPU
+	EndGlobalSection
+	GlobalSection(ProjectConfigurationPlatforms) = postSolution
+		{247F0154-3F6B-409B-BEBA-1BF9302D3B9E}.Debug|AnyCPU.ActiveCfg = Debug|Any CPU
+		{247F0154-3F6B-409B-BEBA-1BF9302D3B9E}.Debug|AnyCPU.Build.0 = Debug|Any CPU
+		{247F0154-3F6B-409B-BEBA-1BF9302D3B9E}.Release|AnyCPU.ActiveCfg = Release|Any CPU
+	EndGlobalSection
+	GlobalSection(SolutionProperties) = preSolution
+		HideSolutionNode = FALSE
+	EndGlobalSection
+EndGlobal

